### PR TITLE
v1.3.6 Improved Session Handling and Logging

### PR DESCRIPTION
**Pull request by kantorcodes**
**Description**:

This PR suggests a number of fixes to improve the dApp experience: 

1. Enables dApps to specify a log level for better control of console logs.
2. Ensures that there is only one Signer per AccountId + ExtensionId combination. Having multiple signers with different ids created issues in pulling the correct record. 
3. Explores better state management and automatic removal of stale sessions. 

**Related issue(s)**:

Fixes CON-494

**Notes for reviewer**:
- We need to merge https://github.com/hashgraph/hedera-wallet-connect/pull/317 first and then update this branch before merging this in.
- These updates are substantial and should be a net positive on the experience, but require careful review from the dApp experience. Wallet experiences should remain mostly the same. 

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

**Review by jwtong — APPROVED** [review]

File: src/lib/dapp/index.ts [hunk 1]
```diff
@@ -112,31 +116,45 @@ export class DAppConnector {
...
+      this.walletConnectClient = signClient
       const existingSessions = this.walletConnectClient.session.getAll()
 
-      if (existingSessions.length > 0)
+      if (existingSessions.length > 0) {
         this.signers = existingSessions.flatMap((session) => this.createSigners(session))
-      else this.checkIframeConnect()
+      } else {
+        await this.checkIframeConnect()
+      }
```
Comment (franfernandez20) [inline src/lib/dapp/index.ts:152]: I agree with improving code style and formatting, but these changes should be in a separate commit to avoid any confusion with the purpose of this one.

File: src/lib/shared/logger.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+      console.info(`[INFO] ${message}`, ...args)
+    }
+  }
+
+  debug(message: string, ...args: any[]): void {
+    if (this.logLevel === 'debug') {
+      console.debug(`[DEBUG] ${message}`, ...args)
+    }
+  }
+}
```
Comment (franfernandez20) [inline src/lib/shared/logger.ts:42]: This is a great idea! However, I'd suggest keeping the logging feature separate from the wallet connect functionality. Putting it in its own commit will keep things clear.

File: src/lib/shared/utils.ts [hunk 3]
```diff
@@ -104,11 +104,11 @@ export function transactionToTransactionBody<T extends Transaction>(
...
   //@ts-ignore
-  return transaction._signedTransactions.current.bodyBytes
+  return transaction._makeTransactionBody(nodeAccountId)
 }
 
-export function transactionBodyToBase64String(transactionBody: Uint8Array) {
-  return Uint8ArrayToBase64String(transactionBody)
+export function transactionBodyToBase64String(transactionBody: proto.ITransactionBody) {
+  return Uint8ArrayToBase64String(proto.TransactionBody.encode(transactionBody).finish())
 }
```
Comment (franfernandez20) [inline src/lib/shared/utils.ts:112]: Same here, this fixes a reported issue, so it should be committed separately and liked to it.

**Comment by franfernandez20** [comment]
Great work, @kantorcodes ! This is a fantastic improvement. The only thing I’d suggest is splitting these changes into four parts: 
- Logging System Implementation
- Session Update Enhancements
- Issue with `transactionBodyToBase64String` Resolution
- Code Style Improvements

**Comment by kantorcodes** [comment]
Thank you for your feedback.

While I generally agree that breaking this out into multiple commits would make this easier in case we must revert something, at the same time the work here provides essential fixes to make HWC more usable for important dApps in the ecosystem. 

Perhaps prioritizing canary releases would be a good interim solution, as I fear the time to review four separate pull requests would slow things down immensely. 

Additionally, while logging is not a required dependency here, it did make developing and debugging these improvements easier.

My proposed options in order:

A) prioritize canary NPM tags so we can allow immediate usage of these improvements to dApps 

B) Push this one over in light of urgency

C) Counter recommendations for two PRs instead of 4

- remove style improvements 
- combine session improvements + logging
- issue a revert PR on transaction bytes separately

**Comment by itsbrandondev** [comment]
@kantorcodes i really like the idea of utilizing canary releases (option A). we have to find a balance between urgent response for dApps and plenty of time for reviews and testing. option C is reasonable as well.

interested to hear thoughts from @teacoat @valeraOlexienko @jwtong

**Comment by kantorcodes** [comment]
We've broken this PR into three new reviews

https://github.com/hashgraph/hedera-wallet-connect/pull/327
https://github.com/hashgraph/hedera-wallet-connect/pull/326
https://github.com/hashgraph/hedera-wallet-connect/pull/325

**Review by andrewb1269hg — COMMENTED** [review]
@rbarkerSL I think these 3 files look good, but please review as well:
`.github/workflows/publish-canary.yml` - new file, definitely review.
`package-lock.json` (doesn't really need to be reviewed)
`package.json`
